common/djangoapps/student/models.py

```diff
@@ -1030,16 +1030,32 @@ def get_or_create_enrollment(cls, user, course_key):
         if user.id is None:
             user.save()
 
-        enrollment, created = cls.objects.get_or_create(
-            user=user,
-            course_id=course_key,
-        )
+        try:
+            enrollment, created = cls.objects.get_or_create(
+                user=user,
+                course_id=course_key,
+            )
 
-        # If we *did* just create a new enrollment, set some defaults
-        if created:
-            enrollment.mode = CourseMode.DEFAULT_MODE_SLUG
-            enrollment.is_active = False
-            enrollment.save()
+            # If we *did* just create a new enrollment, set some defaults
+            if created:
+                enrollment.mode = CourseMode.DEFAULT_MODE_SLUG
+                enrollment.is_active = False
+                enrollment.save()
+
+        except IntegrityError:
+            log.info(
+                (
+                    "An integrity error occurred while getting-or-creating the enrollment"
+                    "for course key %s and student %s. This can occur if two processes try to get-or-create "
+                    "the enrollment at the same time and the database is set to REPEATABLE READ. We will try "
+                    "committing the transaction and retrying."
+                ),
+                course_key, user
+            )
+            enrollment = cls.objects.get(
```

Comment:
I second the recommendation to use the `defaults` kwargs, rather than updating the enrollment after it has been created and returned.